File: 2306-CreateBinaryTreeFromDescriptions/2306-CreateBinaryTreeFromDescriptions.py
# Last updated: 4/29/2026, 10:21:11 AM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        from collections import defaultdict
        map = defaultdict(lambda: None)
        
        for parent, child, isLeft in descriptions:
            parent_node, child_node = None, None
            if map[parent] == None:
                parent_node = TreeNode(parent, None, None)
                map[parent] = [parent_node, False]
            else:
                parent_node = map[parent][0]
            if map[child] == None:
                child_node = TreeNode(child, None, None)
                map[child] = [child_node, True]
            else:
                child_node = map[child][0]
                map[child][1] = True
            if isLeft: 
                parent_node.left = child_node
            else:
                parent_node.right = child_node
        
        for k, v in map.items():
            node, is_child = v
            if not is_child:
                return node

File: 2306-CreateBinaryTreeFromDescriptions/2306-CreateBinaryTreeFromDescriptions.js
// Last updated: 4/29/2026, 10:21:18 AM
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[][]} descriptions
 * @return {TreeNode}
 */
// 23:10s
// O(n)
// O(n)
var createBinaryTree = function(descriptions) {
    const map = {};
    for(const [parent, child, isLeft] of descriptions) {
        // console.log(parent, child, isLeft);
        let parentNode, childNode;
        if(map[parent] === undefined) {
            parentNode = new TreeNode(parent, null, null);
            map[parent] = [parentNode, false];
        } else {
            parentNode = map[parent][0];
        }
        if(map[child] === undefined) {
            childNode = new TreeNode(child, null,null);
            map[child] = [childNode, true];
        } else {
            childNode = map[child][0];
            map[child][1] = true;
        }
        if(isLeft) {
            parentNode.left = childNode;
        } else {
            parentNode.right = childNode;
        }
    }
    // console.log(map);
    const keys = Object.keys(map);
    for(const key of keys) {
        if(!map[key][1]) {
            return map[key][0];
        }
    }
};